typing/typecore.ml

```diff
@@ -2655,6 +2655,18 @@ let check_partial_application ~statement exp =
   | _ ->
       doit ()
 
+let pattern_needs_partial_application_check p =
+  let rec check : type a. a general_pattern -> bool = fun p ->
+    match p.pat_desc with
+    | Tpat_any ->
+        List.for_all (function (Tpat_constraint _, _, _) -> false | _ -> true) p.pat_extra
+    | Tpat_exception _ -> true
+    | Tpat_or (p1, p2, _) -> check p1 && check p2
+    | Tpat_value p -> check (p :> value general_pattern)
+    | _ -> false
+  in
```

Comment:
I think that it is slightly wrong that only `Tpat_any` checks the p.pat_extra. For example I would expect `((_ | _) : foo -> bar)` to disable the warning. I would expect the code to look at both pat_extra and pat_desc independently and perform a conjunction of the results.

Reply:
Yes, it is somewhat arbitrary; it was the existing logic in the `let _` case. I pushed a commit with your suggestion.